experimental/web/sample_webgpu/main.c

```diff
@@ -279,8 +293,192 @@ static iree_status_t parse_inputs_into_call(
   return iree_ok_status();
 }
 
+typedef struct iree_buffer_map_userdata_t {
+  iree_hal_buffer_view_t* buffer_view;
+  iree_hal_buffer_t* mappable_buffer;
+  iree_hal_allocator_t* heap_allocator;
+} iree_buffer_map_userdata_t;
+
+// TODO(scotttodd): move async mapping into webgpu/buffer.h/.c?
+static void buffer_map_sync_callback(WGPUBufferMapAsyncStatus map_status,
+                                     void* userdata_ptr) {
+  iree_buffer_map_userdata_t* userdata =
+      (iree_buffer_map_userdata_t*)userdata_ptr;
+  switch (map_status) {
+    case WGPUBufferMapAsyncStatus_Success:
+      break;
+    case WGPUBufferMapAsyncStatus_Error:
+      fprintf(stderr, "  buffer_map_sync_callback status: Error\n");
+      break;
+    case WGPUBufferMapAsyncStatus_DeviceLost:
+      fprintf(stderr, "  buffer_map_sync_callback status: DeviceLost\n");
+      break;
+    case WGPUBufferMapAsyncStatus_Unknown:
+    default:
+      fprintf(stderr, "  buffer_map_sync_callback status: Unknown\n");
+      break;
+  }
+
+  if (map_status != WGPUBufferMapAsyncStatus_Success) {
+    iree_allocator_free(iree_allocator_system(), userdata);
+    return;
+  }
+
+  iree_status_t status = iree_ok_status();
+
+  // TODO(scotttodd): bubble result(s) up to the caller (async + callback API)
+
+  iree_device_size_t data_length =
+      iree_hal_buffer_view_byte_length(userdata->buffer_view);
+  WGPUBuffer buffer_handle =
+      iree_hal_webgpu_buffer_handle(userdata->mappable_buffer);
+
+  // For this sample we want to print arbitrary buffers, which is easiest
+  // using the |iree_hal_buffer_view_format| function. Internally, that
+  // function requires synchronous buffer mapping, so we'll first wrap the
+  // already (async) mapped GPU memory into a heap buffer. In a less general
+  // application (or one not requiring pretty logging like this), we could
+  // skip a few buffer copies and other data transformations here.
+
+  const void* data_ptr =
+      wgpuBufferGetConstMappedRange(buffer_handle, /*offset=*/0, data_length);
+
+  // Clone into memory that isn't readonly.
```

Comment:
we don't need writeable memory for formatting things and should get rid of this clone - this is the 3rd copy of the data (one for device, one for host, and now one for this)
![image](https://user-images.githubusercontent.com/75337/199623332-37fc0885-3b07-4443-817e-549c2f5b30bf.png)
with iree_hal_heap_buffer_wrap this should just be:
- wgpuBufferGetConstMappedRange
- iree_hal_heap_buffer_wrap
- iree_hal_buffer_view_create
- iree_hal_buffer_view_fprint

Reply:
How should I handle `iree_byte_span_t` vs `iree_const_byte_span_t`? https://github.com/iree-org/iree/blob/a0c6bd575cb011468f5b4aaaa1339ce944bfc558/runtime/src/iree/hal/buffer_heap.c#L34-L38

Can I const cast and rely on `IREE_HAL_MEMORY_ACCESS_READ` for enforcement?

Reply:
yep! #yolo